### DISCUSSION TITLE
How to use Chi with connect-go?

### DISCUSSION THREAD

**Discussion by abitofhelp** [connectrpc/connect-go#738] [Q&A] [answered]
Hello,

I have a simple application based on the ubiquitous Greeter example at grpc's website. After generating, building, and running the server using Chi's mux, curl, grpcurl, and 'buf curl' requests all fail.  However, when I replace Chi's mux with a ServeMux, they work. I'd appreciate any ideas on correcting this error so I can use Chi and its middleware with a Connect app.  I've posted the source at the bottom of the post.

I am using buf v1.32, Go go1.22.3 darwin/amd64, and Chi v5.0.12.

Thank you for your time and suggestions!

Here are the commands and results:

// GRPC
% grpcurl -plaintext -proto proto/abitofhelp/helloworld/v1/helloworld.proto -d '{"name": "Jane"}' localhost:8080 abitofhelp.helloworld.v1.GreeterService/SayHello

ERROR:
  Code: Unimplemented
  Message: unexpected HTTP status code received from server: 404 (Not Found); transport: received unexpected content-type "text/plain; charset=utf-8"

// HTTP
% curl -X POST -H "Content-Type: application/json" -d '{"name": "Jane"}' "http://127.0.0.1:8080/abitofhelp.helloworld.v1.GreeterService/SayHello"                
404 page not found

// CONNECT
% buf curl --http2-prior-knowledge --schema=proto -d '{"name": "Jane"}' "http://127.0.0.1:8080/abitofhelp.helloworld.v1.GreeterService/SayHello" -v              
buf: * Invoking RPC abitofhelp.helloworld.v1.GreeterService.SayHello
buf: * Dialing (tcp) 127.0.0.1:8080...
buf: * Connected to 127.0.0.1:8080
buf: > (#1) POST /abitofhelp.helloworld.v1.GreeterService/SayHello
buf: > (#1) Accept-Encoding: gzip
buf: > (#1) Connect-Protocol-Version: 1
buf: > (#1) Connect-Timeout-Ms: 119990
buf: > (#1) Content-Type: application/proto
buf: > (#1) User-Agent: connect-go/1.16.1 (go1.22.3) buf/1.32.0
buf: > (#1)
buf: } (#1) [6 bytes data]
buf: * (#1) Finished upload
buf: < (#1) HTTP/2.0 404 Not Found
buf: < (#1) Content-Length: 19
buf: < (#1) Content-Type: text/plain; charset=utf-8
buf: < (#1) Date: Mon, 20 May 2024 07:53:29 GMT
buf: < (#1) X-Content-Type-Options: nosniff
buf: < (#1)
buf: { (#1) [19 bytes data]
buf: * (#1) Call complete
{
   "code": "unimplemented",
   "message": "404 Not Found"
}

// SOURCE CODE
```
package main

import (
	"connectrpc.com/connect"
	"context"
	v1 "github.com/abitofhelp/go_connect_bazelmod/gen/abitofhelp/helloworld/v1"
	"github.com/abitofhelp/go_connect_bazelmod/gen/abitofhelp/helloworld/v1/helloworldv1connect"
	"github.com/go-chi/chi/v5"
	"golang.org/x/net/http2"
	"golang.org/x/net/http2/h2c"
	"log"
	"net/http"
)

type GreeterServer struct {
	helloworldv1connect.GreeterServiceHandler
}

func main() {
	mux := chi.NewRouter()
	//mux := http.NewServeMux()

	mux.Handle(helloworldv1connect.NewGreeterServiceHandler(&GreeterServer{}))

	err := http.ListenAndServe(
		"localhost:8080",
		h2c.NewHandler(mux, &http2.Server{}),
	)
	if err != nil {
		panic(err)
	}
}

// SayHello Endpoint
func (ps *GreeterServer) SayHello(
	ctx context.Context,
	req *connect.Request[v1.SayHelloRequest],
) (*connect.Response[v1.SayHelloResponse], error) {
	// connect.Request and connect.Response give you direct access to headers and
	// trailers. No context-based nonsense!
	log.Println(req.Header().Get("Some-Header"))
	res := connect.NewResponse(&v1.SayHelloResponse{
		Message: "howdy " + req.Msg.Name,
	})
	res.Header().Set("Some-Other-Header", "hello!")
	return res, nil
}
```

**Comment by sudorandom** [comment, marked as answer]
This is more a chi question. Chi's mux.Handle doesn't have the same behavior as net/http's mux.Handle. The `net/http` handler treats the first argument as a prefix. Chi treats the first argument as a pattern. Therefore, you can't use the result of `NewGreeterServiceHandler` as the arguments to chi's mux.Handle function. However, chi does have a separate method, `mux.Mount` (https://pkg.go.dev/github.com/go-chi/chi/v5#Mux.Mount) that has a similar behavior to `net/http`'s mux.Handle.

```go
mux.Mount(helloworldv1connect.NewGreeterServiceHandler(&GreeterServer{}))
```

**Reply by abitofhelp** [reply]
Thank you for the help!  I will take a look at the code that you mentioned to understand the differences better.   With go1.22's improved ServeMux, I am finding that I use it more often than Chi's mux since I can use Chi's middleware.
